airbyte-cdk/python/airbyte_cdk/sources/stream_reader/concurrent/partition_generator.py

```diff
@@ -0,0 +1,36 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from concurrent.futures import Future
+from queue import Queue
+from typing import List, Optional
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.stream_reader.concurrent.stream_partition import StreamPartition
+from airbyte_cdk.sources.streams import Stream
+
+
+class PartitionGenerator:
+    def __init__(self, queue: Optional[Queue[StreamPartition]] = None):
+        self._queue = queue if queue else Queue()
+        self._futures: List[Future[None]] = []
+
+    def generate_partitions(self, stream: Stream, sync_mode: SyncMode, cursor_field: Optional[List[str]]) -> None:
+        stream.logger.debug(f"Generating partitions for stream {stream.name}")
+        for partition in stream.generate_partitions(sync_mode=sync_mode, cursor_field=cursor_field):
+            stream_partition = StreamPartition(stream, partition, cursor_field)
+            self._queue.put(stream_partition)
+        stream.logger.debug(f"Done generating partitions for stream {stream.name}")
+
+    def has_next(self) -> bool:
+        return self._queue.qsize() > 0
+
+    def __iter__(self) -> "PartitionGenerator":
+        return self
+
+    def __next__(self) -> StreamPartition:
+        if self._queue.qsize() > 0:
+            return self._queue.get()
+        else:
+            raise StopIteration
```

Comment:
I don't know how async is usually handled in Python but this statement seems odd to me has the iteration might not be completed. [The doc](https://docs.python.org/3/library/exceptions.html#StopIteration) says `Raised by built-in function next() and an iterator's __next__() method to signal that there are no further items produced by the iterator.` which is not exactly the case because there could be more items to be produces.

I see concepts of [async iterator](https://docs.python.org/3/glossary.html#term-asynchronous-iterator). I don't know much about it but should we implement this interface and not the sync iterator one?